src/Crypto/Secp256k1.hs

```diff
@@ -702,28 +745,30 @@ keyPairPubKeyXOTweakAdd KeyPair{..} Tweak{..} = unsafePerformIO . evalContT $ do
 
 -- | Compute a schnorr signature using a 'KeyPair'. The @ByteString@ must be 32 bytes long to get a @Just@ out of this
 -- function
-schnorrSign :: KeyPair -> ByteString -> Maybe Signature
+schnorrSign :: KeyPair -> ByteString -> IO (Maybe SchnorrSignature)
```

Comment:
Sorry for running you in circles, I think you should break this up into three functions:

```haskell
schnorrSign :: Maybe StdGen -> KeyPair -> ByteString -> Maybe SchnorrSignature
schnorrSign = ...

schnorrSignDeterministic :: KeyPair -> ByteString -> Maybe SchnorrSignature
schnorrSignDeterministic = schnorrSign Nothing

schnorrSignNondeterministic :: KeyPair -> ByteString -> IO (Maybe SchnorrSignature)
schnorrSignNondeterministic kp bs = newStdGen >>= \gen -> schnorrSign (Just gen) kp bs
```